### PR TITLE
optimized isEqual function, 30x faster on integers

### PR DESCRIPTION
**Pull request by gaperton**
and 1,5 times faster than lodash.isEqual
http://jsperf.com/underscore-isequal/3

### CODE REVIEW

**Comment by gaperton** [comment]
Some problems in unit tests. Next time then.

File: underscore.js [hunk 1]
```diff
@@ -1203,7 +1198,13 @@
...
   // Perform a deep comparison to check if two objects are equal.
   _.isEqual = function(a, b) {
-    return eq(a, b);
+      // Identical objects are equal. `0 === -0`, but they aren't identical.
+      // See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
+      if (a === b) return a !== 0 || 1 / a === 1 / b;
+      var typeA = typeof a;
+      if( typeA != typeof b ) return false;
+      if( typeA == 'object' ) return a && b ? eq( a, b ) : a === b;
+      return a != a ? b != b : a === b;
```
Comment (megawac) [inline underscore.js]: You can't move this code so `[NaN] == [NaN]`

File: underscore.js [hunk 2]
```diff
@@ -1110,11 +1110,6 @@
 
   // Internal recursive comparison function for `isEqual`.
   var eq = function(a, b, aStack, bStack) {
-    // Identical objects are equal. `0 === -0`, but they aren't identical.
-    // See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
-    if (a === b) return a !== 0 || 1 / a === 1 / b;
-    // A strict comparison is necessary because `null == undefined`.
-    if (a == null || b == null) return a === b;
```
Comment (megawac) [inline underscore.js]: Sorry I meant this code

File: underscore.js [hunk 3]
```diff
@@ -1203,7 +1198,13 @@
 
   // Perform a deep comparison to check if two objects are equal.
   _.isEqual = function(a, b) {
-    return eq(a, b);
+      // Identical objects are equal. `0 === -0`, but they aren't identical.
+      // See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
+      if (a === b) return a !== 0 || 1 / a === 1 / b;
+      var typeA = typeof a;
+      if( typeA != typeof b ) return false;
```
Comment (jridgewell) [inline underscore.js]: Unfortunately, types aren't enough since `_.isEqual('test', new String('test')) === true`.

**Comment by gaperton** [comment]
Unfortunately, this code doesn't work at all breaking underscore unit tests, so I've prepared different implementation. Which appears to be correct, and it's faster.

https://github.com/jashkenas/underscore/pull/2180

**Pull request closed without merging**